### PR TITLE
[GATE] Refactoring system_token workspace and adding all_workspace

### PR DESCRIPTION
**Pull request by pedroborgeszup**

### CODE REVIEW

File: gate/internal/repository/models/workspace.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package models
+
+import (
+	"github.com/google/uuid"
+	"time"
+)
+
+type Workspace struct {
+	ID                      uuid.UUID
```
Comment (monicaribeiro) [inline gate/internal/repository/models/workspace.go:9]: Why this id is uppercase?

File: gate/internal/repository/models/user_group.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package models
+
+import "time"
+
+type UserGroup struct {
+	id        string
```
Comment (monicaribeiro) [inline gate/internal/repository/models/user_group.go:6]: Put the attributes on cammel case.

File: gate/internal/repository/workspace_repository.go [hunk 3]
```diff
@@ -28,8 +28,9 @@ import (
 )
 
 type WorkspaceRepository interface {
-	CountByIds(workspaceIds []string) (int64, error)
+	FindByIds(workspaceIds []string) ([]domain.SimpleWorkspace, error)
 	GetUserPermissionAtWorkspace(workspaceId string, userId string) ([][]domain.Permission, error)
+	FindWorkspacesBySystemTokenId(systemTokenId string) ([]domain.SimpleWorkspace, error)
```
Comment (monicaribeiro) [inline gate/internal/repository/workspace_repository.go:33]:
```suggestion
	FindBySystemTokenId(systemTokenId string) ([]domain.SimpleWorkspace, error)
```

File: gate/internal/repository/workspace_repository.go [hunk 4]
```diff
@@ -78,6 +79,18 @@ func (workspaceRepository workspaceRepository) GetUserPermissionAtWorkspace(work
 	return permissionsListDomain, nil
 }
 
+func (workspaceRepository workspaceRepository) FindWorkspacesBySystemTokenId(systemTokenId string) ([]domain.SimpleWorkspace, error) {
+	var workspaces []models.Workspace
+
+	res := workspaceRepository.db.Raw(findWorkspacesBySystemTokenIdQuery, systemTokenId).Scan(&workspaces)
+
+	if res.Error != nil {
+		return []domain.SimpleWorkspace{}, handlePermissionError("Find all workspaces failed", "PermissionRepository.FindWorkspacesBySystemTokenId.Find", res.Error, logging.InternalError)
```
Comment (monicaribeiro) [inline gate/internal/repository/workspace_repository.go:88]:
```suggestion
		return []domain.SimpleWorkspace{}, handlePermissionError("Find workspaces by system token id failed.", "PermissionRepository.FindBySystemTokenId.Find", res.Error, logging.InternalError)
```

File: gate/internal/repository/workspace_repository.go [hunk 5]
```diff
@@ -90,3 +103,16 @@ const findUserPermissionsAtWorkspaceQuery = `
 		inner join user_groups_users ugu on ugu.user_group_id = ug.id 
 		inner join users u on ? = ugu.user_id
 `
+
+const findWorkspacesBySystemTokenIdQuery = `
+	select
```
Comment (monicaribeiro) [inline gate/internal/repository/workspace_repository.go:108]:
```suggestion
	SELECT
		id,
		name,
		created_at
	FROM
		workspaces w
	INNER JOIN system_tokens_workspaces stw ON
		p.id = stw.workspace_id
	WHERE
		stw .system_token_id = ?
```

File: gate/internal/use_case/authorization/authorize_system_token.go [hunk 6]
```diff
@@ -62,11 +66,18 @@ func (authorizeSystemToken authorizeSystemToken) Execute(authorizationToken stri
...
+			return logging.WithOperation(err, "authorize.systemToken")
+		}
+
+		if !contains(workspaces, workspaceId) {
+			return logging.NewError("Forbidden", errors.New("forbidden"), logging.ForbiddenError, nil, "authorize.systemToken")
+		}
 	}
 
-	permissions, err := authorizeSystemToken.permissionRepository.FindBySystemTokenId(systemToken.ID.String())
+	permissions, err := authorizeSystemToken.permissionRepository.FindPermissionsBySystemTokenId(systemToken.ID.String())
```
Comment (monicaribeiro) [inline gate/internal/use_case/authorization/authorize_system_token.go:80]:
```suggestion
	permissions, err := authorizeSystemToken.permissionRepository.FindBySystemTokenId(systemToken.ID.String())
```

File: gate/internal/use_case/system_token/create_system_token.go [hunk 7]
```diff
@@ -74,22 +74,23 @@ func (createSystemToken createSystemToken) Execute(authorization string, input C
...
 	}
 
-	workspacesFound, err := createSystemToken.workspaceRepository.CountByIds(input.Workspaces)
+	workspaces, err := createSystemToken.workspaceRepository.FindByIds(input.Workspaces)
 	if err != nil {
 		return domain.SystemToken{}, logging.WithOperation(err, "CreateSystemToken.Execute")
 	}
 
-	if int(workspacesFound) < len(input.Workspaces) {
+	if len(workspaces) < len(input.Workspaces) {
```
Comment (monicaribeiro) [inline gate/internal/use_case/system_token/create_system_token.go:82]: Maybe we could compare the ids too.

File: gate/internal/use_case/system_token/create_system_token_input.go [hunk 8]
```diff
@@ -25,23 +25,25 @@ import (
...
-		Name:        input.Name,
-		Revoked:     false,
-		Permissions: []domain.Permission{},
-		Workspaces:  input.Workspaces,
-		Token:       "",
-		CreatedAt:   &createdAt,
-		RevokedAt:   nil,
-		LastUsedAt:  nil,
-		Author:      "",
+		ID:            uuid.New(),
```
Comment (monicaribeiro) [inline gate/internal/use_case/system_token/create_system_token_input.go:37]: We could put all the ids in cammel case.

File: gate/tests/unit/utils/dummy_utils.go [hunk 9]
```diff
@@ -43,6 +43,19 @@ func GetDummySystemToken() domain.SystemToken {
...
 
+func GetDummySimpleWorkspaces() []domain.SimpleWorkspace {
+	return []domain.SimpleWorkspace{
+		{
+			ID:   uuid.New(),
+			Name: "Workspace Name",
+		},
+		{
+			ID:   uuid.New(),
+			Name: "Workspace Name",
```
Comment (monicaribeiro) [inline gate/tests/unit/utils/dummy_utils.go:54]:
```suggestion
			Name: "Workspace Name 2",
```

Comment (monicaribeiro) [reply gate/tests/unit/utils/dummy_utils.go:54]: A different name can help sometimes in unit tests.

**Review by monicaribeiro — CHANGES_REQUESTED** [review]

File: gate/internal/repository/models/workspace.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+	Name                    string
+	Author                  uuid.UUID `gorm:"column:user_id"`
+	CreatedAt               *time.Time
+	UserGroups              []UserGroup `gorm:"many2many:workspaces_user_groups;"`
+	Status                  string
+	RegistryConfigurationId string
```
Comment (barbararochazup) [inline gate/internal/repository/models/workspace.go:15]: Ids (GitConfigurationId, CdConfigurationId, etc) -> Should it be uuid type?

File: gate/internal/use_case/system_token/create_system_token_input.go (hunk 8 above)
Comment (barbararochazup) [reply gate/internal/use_case/system_token/create_system_token_input.go:37]: Its a good practice (convention). See "Initialisms" topics https://github.com/golang/go/wiki/CodeReviewComments

File: gate/resources/migrations/20210409114800_system_token_workspaces_table.down.sql [hunk 11]
```diff
@@ -0,0 +1,7 @@
+-- ALTER TABLE system_tokens
+--     ADD workspaces VARCHAR(36)[] NOT NULL default [];
+--
+-- ALTER TABLE system_tokens
+--     DROP COLUMN all_workspaces;
```
Comment (barbararochazup) [inline gate/resources/migrations/20210409114800_system_token_workspaces_table.down.sql:5]: Should be commented?

File: gate/internal/repository/models/user_group.go (hunk 2 above)
Comment (monicaribeiro) [reply gate/internal/repository/models/user_group.go]: Ignore this commentttt!! lol https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Comment (monicaribeiro) [reply gate/internal/repository/models/user_group.go]: Just put it as "ID"

File: gate/internal/repository/models/workspace.go (hunk 1 above)
Comment (monicaribeiro) [reply gate/internal/repository/models/workspace.go]: Answer https://github.com/golang/go/wiki/CodeReviewComments#initialisms

**Review by barbararochazup — APPROVED** [review]

**Review by monicaribeiro — APPROVED** [review]

**Pull request merged**